### DISCUSSION TITLE
I broke netbox 502 Bad Gateway

### DISCUSSION THREAD

**Discussion by RegisTB** [netbox-community/netbox#12804] [General]
I had a running env with no issue and wanted to upgrade the box.
I ran the ./upgrade.sh without being cautious of all the red flag behing that action 
what is the proecess to get back to where I was 

I broke -- netbox 502 Bad Gateway

**Comment by martinum4** [comment]
Revert to last Backup? I just have a cronjob in the background that does an DB-Backup every 15 min on the workdays…
Maybe enable debug, check all your logs, restart the daemon

**Comment by RegisTB** [comment]
I am not so savvy at the development environment
But I am finding source to help me out
I cN see the old network environment folder on my server
I need to find a way the restore it or install the new nebox I just pushed

**Reply by martinum4** [reply]
Did you follow the upgrade guide at https://docs.netbox.dev/en/stable/installation/upgrading/ ? Or just run the upgrade.sh without any of the steps? Helping gets kinda difficult when you don't know what has and hasn't been done…

**Comment by RegisTB** [comment]
Yes I saw the script and ran it
But it was clearly stated to back up and follow the steps
I was getting errors because was not running the right Python and also not
in the right directory

**Comment by candlerb** [comment]
Show the output from:

```
systemctl status netbox
journalctl -n100 --no-pager -u netbox
```

**Comment by RegisTB** [comment]
● netbox.service - NetBox WSGI Service
   Loaded: loaded (/etc/systemd/system/netbox.service; enabled; vendor
preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Sat
2023-06-03 06:17:52 >
     Docs: https://netbox.readthedocs.io/en/stable/
  Process: 273809 ExecStart=/opt/netbox/venv/bin/gunicorn --pid
/var/tmp/netbox.pid --p>
 Main PID: 273809 (code=exited, status=203/EXEC)

 journalctl -n100 --no-pager -u netbox


[root@]# journalctl -n100 --no-pager -u netbox
-- Logs begin at Fri 2023-06-02 11:26:04 CDT, end at Sat 2023-06-03
06:20:05 CDT. --
Jun 03 06:11:49 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:11:49 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:11:49 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:11:49 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:12:19 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:12:19 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2234.
Jun 03 06:12:19 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:12:19 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:12:19 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:12:19 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:12:50 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:12:50 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2235.
Jun 03 06:12:50 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:12:50 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:12:50 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:12:50 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:13:20 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:13:20 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2236.
Jun 03 06:13:20 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:13:20 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:13:20 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:13:20 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:13:50 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:13:50 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2237.
Jun 03 06:13:50 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:13:50 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:13:50 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:13:50 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:14:20 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:14:20 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2238.
Jun 03 06:14:20 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:14:20 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:14:20 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:14:20 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:14:51 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:14:51 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2239.
Jun 03 06:14:51 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:14:51 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:14:51 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:14:51 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:15:21 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:15:21 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2240.
Jun 03 06:15:21 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:15:21 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:15:21 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:15:21 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:15:51 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:15:51 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2241.
Jun 03 06:15:51 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:15:51 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:15:51 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:15:51 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:16:21 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:16:21 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2242.
Jun 03 06:16:21 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:16:21 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:16:21 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:16:21 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:16:52 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:16:52 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2243.
Jun 03 06:16:52 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:16:52 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:16:52 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:16:52 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:17:22 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:17:22 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2244.
Jun 03 06:17:22 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:17:22 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:17:22 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:17:22 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:17:52 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:17:52 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2245.
Jun 03 06:17:52 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:17:52 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:17:52 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:17:52 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:18:22 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:18:22 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2246.
Jun 03 06:18:22 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:18:22 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:18:22 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:18:22 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:18:53 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:18:53 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2247.
Jun 03 06:18:53 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:18:53 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:18:53 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:18:53 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:19:23 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:19:23 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2248.
Jun 03 06:19:23 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:19:23 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:19:23 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:19:23 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.
Jun 03 06:19:53 NBserver systemd[1]: netbox.service: Service RestartSec=30s
expired, scheduling restart.
Jun 03 06:19:53 NBserver systemd[1]: netbox.service: Scheduled restart job,
restart counter is at 2249.
Jun 03 06:19:53 NBserver systemd[1]: Stopped NetBox WSGI Service.
Jun 03 06:19:53 NBserver systemd[1]: Started NetBox WSGI Service.
Jun 03 06:19:53 NBserver systemd[1]: netbox.service: Main process exited,
code=exited, status=203/EXEC
Jun 03 06:19:53 NBserver systemd[1]: netbox.service: Failed with result
'exit-code'.

**Reply by candlerb** [reply]
Error 203 from systemd [means](https://www.freedesktop.org/software/systemd/man/systemd.exec.html): _"The actual process execution failed (specifically, the `execve(2)` system call). Most likely this is caused by a missing or non-accessible executable file."_

Check in `/etc/systemd/system/netbox.service` to see what you have in the "User" and "ExecStart" lines:

```
User=netbox
...
ExecStart=/opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
```

Now try running it by hand:

```
cd /var/tmp
sudo -u netbox /opt/netbox/venv/bin/gunicorn --pid /var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config /opt/netbox/gunicorn.py netbox.wsgi
```

You should see the error.  It could be a permissions error - e.g. the gunicorn script is not executable - or you might have missed one of several installation steps.

Another possible problem is that you've upgraded the underlying OS and now the python version is different. If so, you'll need to rebuild the netbox virtualenv by re-running `/opt/netbox/upgrade.sh`

**Comment by RegisTB** [comment]
Perfect!
As soon as I get to the box I will try those.
The dev team is looking at it but I am a networking guy and wanted to
expand my skill not knowing it break easily specially in production.
I owe that repair to my team.
Tha m you

**Comment by RegisTB** [comment]
when i try running the upgrade I get this error

 File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in
_call_with_frames_removed
  File "/opt/netbox-3.0.10/netbox/netbox/settings.py", line 31, in <module>
    f"NetBox requires Python 3.7 or higher (current: Python
{platform.python_version()})"
RuntimeError: NetBox requires Python 3.7 or higher (current: Python 3.6.8)

however the server has the python in the opt
I am at lost I am sure if the correct python is available in my user
directory I can  upgrade it

**Reply by candlerb** [reply]
> RuntimeError: NetBox requires Python 3.7 or higher (current: Python 3.6.8)

I think that error message is pretty clear!!  Try running `python3 -V` to see what version of python you have installed.

If you believe you have a newer version, then it must be runnable using a different command.  Suppose `python3.8 -V` works, then you can do:

```
PYTHON=python3.8 ./upgrade.sh
```

to tell the upgrade script to use that different version.

**Comment by RegisTB** [comment]
systemctl status netbox
● netbox.service - NetBox WSGI Service
   Loaded: loaded (/etc/systemd/system/netbox.service; enabled; vendor
preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Sat
2023-06-03 20:38:54 CDT; 15s ago
     Docs: https://netbox.readthedocs.io/en/stable/
  Process: 433439 ExecStart=/opt/netbox/venv/bin/gunicorn --pid
/var/tmp/netbox.pid --pythonpath /opt/netbox/netbox --config
/opt/netbox/gunicorn.py netbox.wsgi (code=exited, status=1/FAILURE)
 Main PID: 433439 (code=exited, status=1/FAILURE)

**Comment by RegisTB** [comment]
I will point the upgrade to that version and see what happens.
Thank you for your help

**Comment by RegisTB** [comment]
Hello and thank you for looking into this
The upgrade has been completed using the new version of Python.
However after restarting the service I am still getting the same error
message.

502 Bad Gateway
nginx/1.14.1

**Reply by candlerb** [reply]
That just means Netbox isn't running.  So once again, show  new output from:

```
systemctl status netbox
journalctl -n100 --no-pager -u netbox
```

**Comment by RegisTB** [comment]
Hello,
I involved the dev team and we ran the upgrade with  Python 3.8.12
the upgrade looking for the config file
*” (netbox.configuration) not found*

Applying database migrations (python3 netbox/manage.py migrate)...

Traceback (most recent call last):

  File "/opt/netbox-3.4.7/netbox/netbox/settings.py", line 50, in <module>

    configuration = importlib.import_module(config_path)

  File "/opt/Python-3.8.12/Lib/importlib/__init__.py", line 127, in
import_module

    return _bootstrap._gcd_import(name[level:], package, level)

  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import

  File "<frozen importlib._bootstrap>", line 991, in _find_and_load

  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked

ModuleNotFoundError: No module named 'netbox.configuration'

Please define netbox/netbox/configuration.py per the documentation
I am trying to find that file which I located





***@***.*** /]# find -name "configuration.py"
./usr/local/lib/python3.9/site-packages/pip/_internal/configuration.py
./usr/local/lib/python3.9/site-packages/pip/_internal/commands/configuration.py
./usr/local/lib/python3.8/site-packages/pip/_internal/configuration.py
./usr/local/lib/python3.8/site-packages/pip/_internal/commands/configuration.py
./opt/netbox-3.0.10/netbox/netbox/configuration.py
./opt/netbox-3.4.7/venv/lib/python3.8/site-packages/pip/_internal/configuration.py

./opt/netbox-3.4.7/venv/lib/python3.8/site-packages/pip/_internal/commands/configuration.py



I am  not sure what to do at this point

**Reply by candlerb** [reply]
I think you might have missed some steps from [upgrade instructions](https://docs.netbox.dev/en/stable/installation/upgrading/#option-a-download-a-release). 

*"Copy `local_requirements.txt`, `configuration.py`, and `ldap_config.py` (if present) from the current installation to the new version ... Be sure to replicate your uploaded media as well ... Also make sure to copy or link any custom scripts and reports that you've made ... If you followed the original installation guide to set up gunicorn, be sure to copy its configuration as well"*

**Comment by RegisTB** [comment]
I just reverted the server and will tey again
This time I will be careful  following the steps.
 Thank you

**Comment by RegisTB** [comment]
I gave it a try again and now after following all the steps carefully i ran
into this

  File
"/opt/netbox-3.5.0/venv/lib64/python3.11/site-packages/django/db/backends/postgresql/base.py",
line 215, in get_new_connection
    connection = Database.connect(**conn_params)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File
"/opt/netbox-3.5.0/venv/lib64/python3.11/site-packages/psycopg2/__init__.py",
line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
psycopg2.OperationalError: connection to server at "localhost" (::1), port
5432 failed: Connection refused
        Is the server running on that host and accepting TCP/IP connections?
connection to server at "localhost" (127.0.0.1), port 5432 failed:
Connection refused
        Is the server running on that host and accepting TCP/IP connections?

I tried to update teh postgresql.com file to allow the port but still the
same
i am not able to restart the service

**Reply by candlerb** [reply]
> psycopg2.OperationalError: connection to server at "localhost" (::1), port 5432 failed: Connection refused

Most likely your postgres database is not running, or is running on a different host, or is firewalled off, or has been configured not to accept TCP connections.

Try: `ss -natp | grep 5432` to see if anything is listening on port 5432.

> I tried to update teh postgresql.com file to allow the port but still the same

There is no "postgresql.com" file.  Please say *exactly* what file you edited, and what change you made to it.

Also show what database configuration you put in `/opt/netbox/netbox/netbox/configuration.py` (blank out the password).
